src/core/ddsc/src/dds_rhc_default.c

```diff
@@ -708,13 +816,18 @@ static void free_instance_rhc_free_wrap (void *vnode, void *varg)
 
 static void dds_rhc_default_free (struct dds_rhc_default *rhc)
 {
+  ddsrt_mutex_lock (&rhc->lock);
```

Comment:
Locking & unlocking a mutex protecting an object prior to unconditionally freeing it and without any intervening synchronisation to ensure all threads that might have been blocked on the mutex have left their critical regions prior to that freeing is pretty much always a problem.